WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/Synchronizer/RemoteOrderSynchronizerTests.swift

```diff
@@ -650,6 +917,37 @@ final class RemoteOrderSynchronizerTests: XCTestCase {
         XCTAssertEqual(states, [.syncing(blocking: true), .synced])
     }
 
+    func test_states_are_properly_set_upon_success_order_creation_via_productID_input() {
+        // Given
+        let stores = MockStoresManager(sessionManager: .testingInstance)
+        let synchronizer = RemoteOrderSynchronizer(siteID: sampleSiteID, flow: .creation, stores: stores)
+        stores.whenReceivingAction(ofType: OrderAction.self) { action in
+            switch action {
+            case .createOrder(_, _, let onCompletion):
+                onCompletion(.success(.fake()))
+            default:
+                XCTFail("Unexpected action: \(action)")
+            }
+        }
+
+        // When
+        let states: [OrderSyncState] = waitFor { promise in
+            synchronizer.statePublisher
+                .dropFirst()
+                .collect(2)
+                .sink { states in
+                    promise(states)
+                }
```

Comment:
Since you don't actually need the states here, you could just subscribe to the publisher, then only fulfil the promise when the correct state is received in your block?

```suggestion
        waitFor { promise in
            synchronizer.statePublisher
                .sink { state in
                    guard state == .EXPECTED_STATE_HERE else { return }
                    promise(())
                }
```

Reply:
Thanks! That was my initial approach but I couldn't make it work since we're expecting more than one state to happen (`.syncing(blocking: true)`, then `.synced`), and the promise will be called as soon as we receive the first one, so the stream is disposed just after. 

For now I've added a comment to the code here: 2618b4f2d422c8f3f2f8b25ac33f7b6c2c9aa788 explaining what these are doing, and logged the suggestion on [this issue](https://github.com/woocommerce/woocommerce-ios/issues/9301) to give it another go.